### PR TITLE
CI: flake8-pyi

### PR DESCRIPTION
**Pull request by twoertwein**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -229,3 +229,11 @@ repos:
         entry: python scripts/validate_min_versions_in_sync.py
         language: python
         files: ^(ci/deps/actions-.*-minimum_versions\.yaml|pandas/compat/_optional\.py)$
+    -   id: flake8-pyi
+        name: flake8-pyi
+        entry: flake8 --extend-ignore=E301,E302,E305,E701,E704
```
Comment (twoertwein) [inline .pre-commit-config.yaml:234]: flake8 doesn't like blake's formatting in pyi files

File: setup.cfg [hunk 2]
```diff
@@ -107,9 +107,19 @@ ignore =
...
+    # If test must be a simple comparison against sys.platform or sys.version_info
+    Y002,
+    # Use "_typeshed.Self" instead of class-bound TypeVar
+    Y019,
+    # Docstrings should not be included in stubs
+    Y021,
+    # Use typing_extensions.TypeAlias for type aliases
+    Y026,
+    # Use "collections.abc.*" instead of "typing.*" (PEP 585 syntax)
+    Y027
```
Comment (twoertwein) [inline setup.cfg:122]: Y002: we have one `if TYPE_CHECKING`, without it mypy cannot infer some times (too many cyclic dependencies or does it hide actual errors?)
Y019: pandas does not (explicitly) depend on typeshed/typing_extension
Y021: some files have notes on how the cython and python types relate to each other
Y026: pandas does not (explicitly) depend on typing_extension
Y027: let's wait until we also do that change for the in-line annotations

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -229,3 +229,11 @@ repos:
...
+        types: [pyi]
+        language: python
+        additional_dependencies:
```
Comment (mroeschke) [inline .pre-commit-config.yaml:237]: Could these be added to `flake8_dependencies` in the flake8 block above and then this line changed to `additional_dependencies: *flake8_dependencies`?

Comment (twoertwein) [reply .pre-commit-config.yaml:237]: I first tried to add `flake8-pyi` to all the other flake8 plugins but the issue is that we need different ignores for pyi files than for normal py files.

I think flake8 supports file-specific ignores. If they support something along the lines `*.pyi: E301,E302,E305,E701,E704` that would be the best approach. I'll look into that later today.

**Review by mroeschke — APPROVED** [review]

**Pull request merged**